rmw_connext_shared_cpp/src/create_topic.cpp

```diff
@@ -0,0 +1,76 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_connext_shared_cpp/create_topic.hpp"
+
+#include <cassert>
+#include <mutex>
+
+#include "rmw/error_handling.h"
+#include "rmw/types.h"
+
+#include "rmw_connext_shared_cpp/types.hpp"
+#include "rmw_connext_shared_cpp/ndds_include.hpp"
+
+DDS::Topic *
+rmw_connext_shared_cpp::create_topic(
+  const rmw_node_t * node,
+  const char * topic_name,
+  const char * dds_topic_name,
+  const char * dds_topic_type)
+{
+  // This function is internal, and should be called from functions that already verified
```

Comment:
Should we consider moving the implementation to a private header?

Reply:
It's a bit subtle, but we shouldn't.  This function is in `rmw_connext_shared_cpp` package, but is used by the `rmw_connext_cpp` package.  So it needs to be in a public header to reach across the package boundary.

There's a related question about whether we should move this function into the `rmw_connext_cpp` package, but see my overall review comment for a reason why not.

Reply:
I've documented the preconditions for further clarity https://github.com/ros2/rmw_connext/pull/442/commits/a0c3fe1931b984bf259be58600c4b25217711b15.